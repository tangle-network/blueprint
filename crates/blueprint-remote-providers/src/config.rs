//! Cloud provider configuration types

use serde::{Deserialize, Serialize};

/// Cloud provider configuration for all supported providers
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CloudConfig {
    pub enabled: bool,
    pub aws: Option<AwsConfig>,
    pub gcp: Option<GcpConfig>,
    pub azure: Option<AzureConfig>,
    pub digital_ocean: Option<DigitalOceanConfig>,
    pub vultr: Option<VultrConfig>,
}

/// AWS configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AwsConfig {
    pub enabled: bool,
    pub region: String,
    pub access_key: String,
    pub secret_key: String,
    pub priority: Option<u8>,
}

/// GCP configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GcpConfig {
    pub enabled: bool,
    pub region: String,
    pub project_id: String,
    pub service_account_path: String,
    pub priority: Option<u8>,
}

/// Azure configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AzureConfig {
    pub enabled: bool,
    pub region: String,
    pub client_id: String,
    pub client_secret: String,
    pub tenant_id: String,
    pub priority: Option<u8>,
}

/// DigitalOcean configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DigitalOceanConfig {
    pub enabled: bool,
    pub region: String,
    pub api_token: String,
    pub priority: Option<u8>,
}

/// Vultr configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VultrConfig {
    pub enabled: bool,
    pub region: String,
    pub api_key: String,
    pub priority: Option<u8>,
}

impl CloudConfig {
    /// Load configuration from environment variables
    pub fn from_env() -> Option<Self> {
        use std::env;

        let mut cloud_config = CloudConfig::default();
        let mut any_enabled = false;

        // AWS configuration
        if let (Ok(key), Ok(secret)) = (
            env::var("AWS_ACCESS_KEY_ID"),
            env::var("AWS_SECRET_ACCESS_KEY"),
        ) {
            cloud_config.aws = Some(AwsConfig {
                enabled: true,
                region: env::var("AWS_DEFAULT_REGION").unwrap_or_else(|_| "us-east-1".to_string()),
                access_key: key,
                secret_key: secret,
                priority: Some(10),
            });
            any_enabled = true;
        }

        // GCP configuration
        if let Ok(project_id) = env::var("GCP_PROJECT_ID") {
            let service_account_path = env::var("GOOGLE_APPLICATION_CREDENTIALS")
                .unwrap_or_else(|_| "/etc/gcp/service-account.json".to_string());
            cloud_config.gcp = Some(GcpConfig {
                enabled: true,
                region: env::var("GCP_DEFAULT_REGION")
                    .unwrap_or_else(|_| "us-central1".to_string()),
                project_id,
                service_account_path,
                priority: Some(8),
            });
            any_enabled = true;
        }

        // Azure configuration
        if let (Ok(client_id), Ok(client_secret), Ok(tenant_id)) = (
            env::var("AZURE_CLIENT_ID"),
            env::var("AZURE_CLIENT_SECRET"),
            env::var("AZURE_TENANT_ID"),
        ) {
            cloud_config.azure = Some(AzureConfig {
                enabled: true,
                region: env::var("AZURE_DEFAULT_REGION").unwrap_or_else(|_| "East US".to_string()),
                client_id,
                client_secret,
                tenant_id,
                priority: Some(7),
            });
            any_enabled = true;
        }

        // DigitalOcean configuration
        if let Ok(token) = env::var("DO_API_TOKEN") {
            cloud_config.digital_ocean = Some(DigitalOceanConfig {
                enabled: true,
                region: env::var("DO_DEFAULT_REGION").unwrap_or_else(|_| "nyc3".to_string()),
                api_token: token,
                priority: Some(5),
            });
            any_enabled = true;
        }

        // Vultr configuration
        if let Ok(key) = env::var("VULTR_API_KEY") {
            cloud_config.vultr = Some(VultrConfig {
                enabled: true,
                region: env::var("VULTR_DEFAULT_REGION").unwrap_or_else(|_| "ewr".to_string()),
                api_key: key,
                priority: Some(3),
            });
            any_enabled = true;
        }

        if any_enabled {
            cloud_config.enabled = true;
            Some(cloud_config)
        } else {
            None
        }
    }
}
