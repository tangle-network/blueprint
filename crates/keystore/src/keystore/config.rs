use crate::remote::RemoteConfig;
use std::path::PathBuf;

#[derive(Default)]
pub struct KeystoreConfig {
    pub(crate) in_memory: bool,
    #[cfg(feature = "std")]
    pub(crate) fs_root: Option<PathBuf>,
    #[cfg(any(
        feature = "aws-signer",
        feature = "gcp-signer",
        feature = "ledger-browser",
        feature = "ledger-node"
    ))]
    pub(crate) remote_configs: Vec<RemoteConfig>,
}

impl KeystoreConfig {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn in_memory(mut self, value: bool) -> Self {
        self.in_memory = value;
        self
    }

    #[cfg(feature = "std")]
    pub fn fs_root(mut self, path: PathBuf) -> Self {
        self.fs_root = Some(path);
        self
    }

    cfg_remote! {
        pub fn remote(mut self, remote_config: RemoteConfig) -> Self {
            self.remote_configs.push(remote_config);
            self
        }
    }

    fn is_empty(&self) -> bool {
        let mut is_empty = self.in_memory;
        #[cfg(feature = "std")]
        {
            is_empty |= self.fs_root.is_none();
        }
        #[cfg(any(
            feature = "aws-signer",
            feature = "gcp-signer",
            feature = "ledger-browser",
            feature = "ledger-node"
        ))]
        {
            is_empty |= !self.remote_configs.is_empty();
        }

        is_empty
    }

    pub(crate) fn finalize(self) -> Self {
        if self.is_empty() {
            return self.in_memory(true);
        }

        self
    }
}
