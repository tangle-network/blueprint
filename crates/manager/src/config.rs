use crate::error::{Error, Result};
use clap::Parser;
use docktopus::bollard::system::Version;
use docktopus::bollard::{API_DEFAULT_VERSION, Docker};
use http_body_util::Full;
use hyper::body::Bytes;
use hyper::header::HeaderValue;
use hyper_util::client::legacy::Client;
use hyper_util::rt::TokioExecutor;
use std::fmt::Display;
use std::net::{IpAddr, Ipv4Addr};
use std::path::PathBuf;
use std::sync::LazyLock;
use tracing::error;
use url::Url;

pub static DEFAULT_DOCKER_HOST: LazyLock<Url> =
    LazyLock::new(|| Url::parse("unix:///var/run/docker.sock").unwrap());

#[derive(Debug, Parser)]
#[command(
    name = "Blueprint Manager",
    about = "An program executor that connects to the Tangle network and runs blueprints dynamically on the fly"
)]
pub struct BlueprintManagerConfig {
    /// The path to the blueprint configuration file
    #[arg(short = 's', long)]
    pub blueprint_config: Option<PathBuf>,
    /// The path to the keystore
    #[arg(short = 'k', long)]
    pub keystore_uri: String,
    /// The directory in which all blueprints will store their data
    #[arg(long, short = 'd', default_value = "./data")]
    pub data_dir: PathBuf,
    /// The cache directory for blueprint manager downloads
    #[arg(long, short = 'd', default_value_os_t = default_cache_dir())]
    pub cache_dir: PathBuf,
    /// The verbosity level, can be used multiple times to increase verbosity
    #[arg(long, short = 'v', action = clap::ArgAction::Count)]
    pub verbose: u8,
    /// Whether to use pretty logging
    #[arg(long)]
    pub pretty: bool,
    /// An optional unique string identifier for the blueprint manager to differentiate between multiple
    /// running instances of a `BlueprintManager` (mostly for debugging purposes)
    #[arg(long, alias = "id")]
    pub instance_id: Option<String>,
    #[arg(long, short = 't')]
    pub test_mode: bool,
    /// Whether to allow invalid GitHub attestations (binary integrity checks)
    ///
    /// This will also allow for running the manager without the GitHub CLI installed.
    #[arg(long)]
    pub allow_unchecked_attestations: bool,
    /// The preferred way to run a blueprint.
    ///
    /// This is not a guarantee that the blueprint will use this method, as there may not be a source
    /// available of this type.
    #[arg(long, short, default_value_t)]
    pub preferred_source: SourceType,
    /// The location of the Podman-Docker socket
    #[arg(long, short, default_value_t = DEFAULT_DOCKER_HOST.clone())]
    pub podman_host: Url,
    /// Authentication proxy options
    #[command(flatten)]
    pub auth_proxy_opts: AuthProxyOpts,
}

fn default_cache_dir() -> PathBuf {
    match dirs::cache_dir() {
        Some(dir) => dir.join("blueprint-manager"),
        None => PathBuf::from("./blueprint-manager-cache"),
    }
}

impl Default for BlueprintManagerConfig {
    fn default() -> Self {
        Self {
            blueprint_config: None,
            keystore_uri: "./keystore".into(),
            data_dir: PathBuf::from("./data"),
            cache_dir: default_cache_dir(),
            verbose: 0,
            pretty: false,
            instance_id: None,
            test_mode: false,
            allow_unchecked_attestations: false,
            preferred_source: SourceType::default(),
            podman_host: DEFAULT_DOCKER_HOST.clone(),
            auth_proxy_opts: AuthProxyOpts::default(),
        }
    }
}

#[derive(clap::ValueEnum, Debug, Copy, Clone, Default, PartialEq, Eq)]
pub enum SourceType {
    Container,
    #[default]
    Native,
    Wasm,
}

impl Display for SourceType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SourceType::Container => write!(f, "container"),
            SourceType::Native => write!(f, "native"),
            SourceType::Wasm => write!(f, "wasm"),
        }
    }
}

pub struct SourceCandidates {
    pub container: Option<Url>,
    pub wasm_runtime: Option<String>,
    pub preferred_source: SourceType,
}

impl SourceCandidates {
    /// Determine all runtime sources available on this system
    ///
    /// # Errors
    ///
    /// If the `preferred_source` is not available on this system, or there is an error during its
    /// detection. Errors that occur while searching for non-preferred sources are silently ignored.
    pub async fn load(preferred_source: SourceType, podman_host: Url) -> Result<SourceCandidates> {
        let mut ret = SourceCandidates {
            container: None,
            wasm_runtime: None,
            preferred_source,
        };

        if let Err(e) = ret.determine_podman(podman_host).await {
            if preferred_source == SourceType::Container {
                error!("Podman not found, cannot use container source type as default: {e}");
                return Err(e);
            }
        }

        if let Err(e) = ret.determine_wasm().await {
            if preferred_source == SourceType::Wasm {
                error!("No WASM runtime found, cannot use WASM source type as default: {e}");
                return Err(e);
            }
        }

        Ok(ret)
    }

    async fn determine_podman(&mut self, host: Url) -> Result<()> {
        fn check_server_header(server: Option<&HeaderValue>) -> bool {
            if let Some(server) = server {
                if let Ok(server) = server.to_str() {
                    return server.to_lowercase().contains("libpod");
                }
            }
            false
        }

        let client = Docker::connect_with_local(host.as_str(), 20, API_DEFAULT_VERSION)
            .map_err(|e| Error::Other(e.to_string()))?;

        // Check the version, cheapest route
        let ver: Version = client
            .version()
            .await
            .map_err(|e| Error::Other(format!("Unable to determine the Podman version: {e}")))?;
        if let Some(comps) = &ver.components {
            if comps
                .iter()
                .any(|c| c.name.to_lowercase().contains("podman"))
            {
                self.container = Some(host);
                return Ok(());
            }
        }
        if let Some(platform) = &ver.platform {
            if platform.name.to_lowercase().contains("podman") {
                self.container = Some(host);
                return Ok(());
            }
        }

        // Fallback, read the HTTP "Server" header from a /_ping
        let res = if let Some(socket) = host.as_str().strip_prefix("unix://") {
            let client = Client::builder(TokioExecutor::new())
                .build::<_, Full<Bytes>>(hyperlocal::UnixConnector);
            client
                .get(hyperlocal::Uri::new(socket, "/_ping").into())
                .await
                .map_err(|e| Error::Other(format!("Unable to reach specified Podman host: {e}")))?
        } else {
            let client = Client::builder(TokioExecutor::new()).build_http::<Full<Bytes>>();
            client
                .get(
                    format!("{host}/_ping")
                        .parse()
                        .map_err(|e| Error::Other(format!("Unable to parse provided URI: {e}")))?,
                )
                .await
                .map_err(|e| Error::Other(format!("Unable to reach specified Podman host: {e}")))?
        };

        if check_server_header(res.headers().get("Server")) {
            self.container = Some(host);
            return Ok(());
        }

        Err(Error::Other(String::from("No Podman-Docker socket found")))
    }

    #[expect(clippy::unused_async, reason = "TBD")]
    async fn determine_wasm(&mut self) -> Result<bool> {
        // TODO: Verify WASM runtime installations
        Ok(true)
    }
}

/// The options for the auth proxy
#[derive(Debug, Parser, Clone)]
pub struct AuthProxyOpts {
    /// The host on which the auth proxy will listen
    #[arg(long, default_value = "0.0.0.0")]
    pub auth_proxy_host: IpAddr,
    /// The port on which the auth proxy will listen
    #[arg(long, default_value_t = 8276)]
    pub auth_proxy_port: u16,
}

impl Default for AuthProxyOpts {
    fn default() -> Self {
        Self {
            auth_proxy_host: IpAddr::V4(Ipv4Addr::UNSPECIFIED),
            auth_proxy_port: 8276, // T9 Mapping of TBPM (Tangle Blueprint Manager)
        }
    }
}
