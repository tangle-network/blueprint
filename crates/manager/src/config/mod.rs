use crate::error::{Error, Result};
use blueprint_auth::proxy::DEFAULT_AUTH_PROXY_PORT;
use blueprint_core::{error, info};
use clap::{Args, Parser};
use docktopus::bollard::system::Version;
use docktopus::bollard::{API_DEFAULT_VERSION, Docker};
use http_body_util::Full;
use hyper::body::Bytes;
use hyper::header::HeaderValue;
use hyper_util::client::legacy::Client;
use hyper_util::rt::TokioExecutor;
use std::fmt::Display;
use std::net::{IpAddr, Ipv4Addr};
use std::path::{Path, PathBuf};
use std::sync::LazyLock;
use url::Url;

pub static DEFAULT_DOCKER_HOST: LazyLock<Url> =
    LazyLock::new(|| Url::parse("unix:///var/run/docker.sock").unwrap());

#[cfg(feature = "vm-sandbox")]
pub static DEFAULT_ADDRESS_POOL: LazyLock<ipnet::Ipv4Net> =
    LazyLock::new(|| "172.30.0.0/16".parse().unwrap());

#[derive(Debug, Parser)]
#[command(
    name = "Blueprint Manager",
    about = "A program executor that connects to the Tangle network and runs blueprints dynamically on the fly"
)]
pub struct BlueprintManagerCli {
    #[command(flatten)]
    pub config: BlueprintManagerConfig,
}

#[derive(Debug, Args)]
pub struct BlueprintManagerConfig {
    /// The path to the blueprint configuration file
    #[arg(short = 'c', long)]
    pub blueprint_config: Option<PathBuf>,
    /// The path to the keystore
    #[arg(short = 'k', long, default_value = "./keystore")]
    pub keystore_uri: String,
    /// The directory in which all blueprints will store their data
    #[arg(long, short = 'd', default_value = "./data")]
    pub data_dir: PathBuf,
    /// The cache directory for blueprint manager downloads
    #[arg(long, short = 'z', default_value_os_t = default_cache_dir())]
    pub cache_dir: PathBuf,
    /// The runtime directory for manager-to-blueprint sockets
    #[arg(long, short, default_value_os_t = default_runtime_dir())]
    pub runtime_dir: PathBuf,
    /// The verbosity level, can be used multiple times to increase verbosity
    #[arg(long, short = 'v', action = clap::ArgAction::Count)]
    pub verbose: u8,
    /// Whether to use pretty logging
    #[arg(long)]
    pub pretty: bool,
    /// An optional unique string identifier for the blueprint manager to differentiate between multiple
    /// running instances of a `BlueprintManager` (mostly for debugging purposes)
    #[arg(long, alias = "id")]
    pub instance_id: Option<String>,
    #[arg(long, short = 't')]
    pub test_mode: bool,
    /// Disables the VM sandbox for native blueprints
    ///
    /// This should only be used for testing and never for production setups.
    #[arg(long)]
    #[cfg(feature = "vm-sandbox")]
    pub no_vm: bool,
    /// Whether to allow invalid GitHub attestations (binary integrity checks)
    ///
    /// This will also allow for running the manager without the GitHub CLI installed.
    #[arg(long)]
    pub allow_unchecked_attestations: bool,
    /// The preferred way to run a blueprint.
    ///
    /// This is not a guarantee that the blueprint will use this method, as there may not be a source
    /// available of this type.
    #[arg(long, short = 's', default_value_t)]
    pub preferred_source: SourceType,
    /// The location of the Podman-Docker socket
    #[arg(long, default_value_t = DEFAULT_DOCKER_HOST.clone())]
    pub podman_host: Url,
    /// The default address pool for VM TAP interfaces
    #[arg(long, default_value_t = *DEFAULT_ADDRESS_POOL)]
    #[cfg(feature = "vm-sandbox")]
    pub default_address_pool: ipnet::Ipv4Net,
    /// The network interface to funnel blueprint VM traffic through
    #[arg(long)]
    #[cfg(feature = "vm-sandbox")]
    pub network_interface: Option<String>,

    /// Authentication proxy options
    #[command(flatten)]
    pub auth_proxy_opts: AuthProxyOpts,
}

impl BlueprintManagerConfig {
    /// Check if all configured directories exist, and if not, create them
    ///
    /// # Errors
    ///
    /// This will error if it fails to create any of the directories.
    pub fn verify_directories_exist(&self) -> Result<()> {
        if !self.cache_dir.exists() {
            info!(
                "Cache directory does not exist, creating it at `{}`",
                self.cache_dir.display()
            );
            std::fs::create_dir_all(&self.cache_dir)?;
        }

        if !self.runtime_dir.exists() {
            info!(
                "Runtime directory does not exist, creating it at `{}`",
                self.runtime_dir.display()
            );
            std::fs::create_dir_all(&self.runtime_dir)?;
        }

        if !self.data_dir.exists() {
            info!(
                "Data directory does not exist, creating it at `{}`",
                self.data_dir.display()
            );
            std::fs::create_dir_all(&self.data_dir)?;
        }

        let keystore = Path::new(&self.keystore_uri);
        if !keystore.exists() {
            info!(
                "Keystore directory does not exist, creating it at `{}`",
                keystore.display()
            );
            std::fs::create_dir_all(keystore)?;
        }

        Ok(())
    }

    /// Checks if a network interface was provided, and if not, attempts to determine the host's default
    ///
    /// # Errors
    ///
    /// This will error if it is unable to determine the default network interface
    #[cfg(feature = "vm-sandbox")]
    pub fn verify_network_interface(&mut self) -> Result<()> {
        if self.network_interface.is_some() {
            return Ok(());
        }

        let Ok(interface) = netdev::interface::get_default_interface().map(|i| i.name) else {
            error!(
                "Unable to determine the default network interface, you must specify it manually with --network-interface"
            );
            return Err(Error::Other(String::from(
                "Failed to determine default network interface",
            )));
        };

        self.network_interface = Some(interface);
        Ok(())
    }
}

fn default_cache_dir() -> PathBuf {
    match dirs::cache_dir() {
        Some(dir) => dir.join("blueprint-manager"),
        None => PathBuf::from("./blueprint-manager-cache"),
    }
}

fn default_runtime_dir() -> PathBuf {
    match dirs::runtime_dir() {
        Some(dir) => dir.join("blueprint-manager"),
        None => PathBuf::from("/run/blueprint-manager"),
    }
}

impl Default for BlueprintManagerConfig {
    fn default() -> Self {
        Self {
            blueprint_config: None,
            keystore_uri: "./keystore".into(),
            data_dir: PathBuf::from("./data"),
            cache_dir: default_cache_dir(),
            runtime_dir: default_runtime_dir(),
            verbose: 0,
            pretty: false,
            instance_id: None,
            test_mode: false,
            #[cfg(feature = "vm-sandbox")]
            no_vm: false,
            allow_unchecked_attestations: false,
            preferred_source: SourceType::default(),
            podman_host: DEFAULT_DOCKER_HOST.clone(),
            #[cfg(feature = "vm-sandbox")]
            default_address_pool: *DEFAULT_ADDRESS_POOL,
            #[cfg(feature = "vm-sandbox")]
            network_interface: None,
            auth_proxy_opts: AuthProxyOpts::default(),
        }
    }
}

#[derive(clap::ValueEnum, Debug, Copy, Clone, Default, PartialEq, Eq)]
pub enum SourceType {
    Container,
    #[default]
    Native,
    Wasm,
}

impl Display for SourceType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SourceType::Container => write!(f, "container"),
            SourceType::Native => write!(f, "native"),
            SourceType::Wasm => write!(f, "wasm"),
        }
    }
}

pub struct SourceCandidates {
    pub container: Option<Url>,
    pub wasm_runtime: Option<String>,
    pub preferred_source: SourceType,
}

impl SourceCandidates {
    /// Determine all runtime sources available on this system
    ///
    /// # Errors
    ///
    /// If the `preferred_source` is not available on this system, or there is an error during its
    /// detection. Errors that occur while searching for non-preferred sources are silently ignored.
    pub async fn load(preferred_source: SourceType, podman_host: Url) -> Result<SourceCandidates> {
        let mut ret = SourceCandidates {
            container: None,
            wasm_runtime: None,
            preferred_source,
        };

        if let Err(e) = ret.determine_podman(podman_host).await {
            if preferred_source == SourceType::Container {
                error!("Podman not found, cannot use container source type as default: {e}");
                return Err(e);
            }
        }

        if let Err(e) = ret.determine_wasm().await {
            if preferred_source == SourceType::Wasm {
                error!("No WASM runtime found, cannot use WASM source type as default: {e}");
                return Err(e);
            }
        }

        Ok(ret)
    }

    async fn determine_podman(&mut self, host: Url) -> Result<()> {
        fn check_server_header(server: Option<&HeaderValue>) -> bool {
            if let Some(server) = server {
                if let Ok(server) = server.to_str() {
                    return server.to_lowercase().contains("libpod");
                }
            }
            false
        }

        let client = Docker::connect_with_local(host.as_str(), 20, API_DEFAULT_VERSION)
            .map_err(|e| Error::Other(e.to_string()))?;

        // Check the version, cheapest route
        let ver: Version = client
            .version()
            .await
            .map_err(|e| Error::Other(format!("Unable to determine the Podman version: {e}")))?;
        if let Some(comps) = &ver.components {
            if comps
                .iter()
                .any(|c| c.name.to_lowercase().contains("podman"))
            {
                self.container = Some(host);
                return Ok(());
            }
        }
        if let Some(platform) = &ver.platform {
            if platform.name.to_lowercase().contains("podman") {
                self.container = Some(host);
                return Ok(());
            }
        }

        // Fallback, read the HTTP "Server" header from a /_ping
        let res = if let Some(socket) = host.as_str().strip_prefix("unix://") {
            let client = Client::builder(TokioExecutor::new())
                .build::<_, Full<Bytes>>(hyperlocal::UnixConnector);
            client
                .get(hyperlocal::Uri::new(socket, "/_ping").into())
                .await
                .map_err(|e| Error::Other(format!("Unable to reach specified Podman host: {e}")))?
        } else {
            let client = Client::builder(TokioExecutor::new()).build_http::<Full<Bytes>>();
            client
                .get(
                    format!("{host}/_ping")
                        .parse()
                        .map_err(|e| Error::Other(format!("Unable to parse provided URI: {e}")))?,
                )
                .await
                .map_err(|e| Error::Other(format!("Unable to reach specified Podman host: {e}")))?
        };

        if check_server_header(res.headers().get("Server")) {
            self.container = Some(host);
            return Ok(());
        }

        Err(Error::Other(String::from("No Podman-Docker socket found")))
    }

    #[expect(clippy::unused_async, reason = "TBD")]
    async fn determine_wasm(&mut self) -> Result<bool> {
        // TODO: Verify WASM runtime installations
        Ok(true)
    }
}

/// The options for the auth proxy
#[derive(Debug, Parser, Clone)]
pub struct AuthProxyOpts {
    /// The host on which the auth proxy will listen
    #[arg(long, default_value = "0.0.0.0")]
    pub auth_proxy_host: IpAddr,
    /// The port on which the auth proxy will listen
    #[arg(long, default_value_t = DEFAULT_AUTH_PROXY_PORT)]
    pub auth_proxy_port: u16,
}

impl Default for AuthProxyOpts {
    fn default() -> Self {
        Self {
            auth_proxy_host: IpAddr::V4(Ipv4Addr::UNSPECIFIED),
            auth_proxy_port: DEFAULT_AUTH_PROXY_PORT,
        }
    }
}
