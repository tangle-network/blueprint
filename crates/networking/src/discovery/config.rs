use super::{
    behaviour::{DerivedDiscoveryBehaviour, DiscoveryBehaviour},
    new_kademlia,
};
use crate::error::Error;
use blueprint_crypto::KeyType;
use libp2p::{
    Multiaddr, PeerId, StreamProtocol, autonat, identify, identity::PublicKey, mdns, relay, upnp,
};
use std::{
    collections::{HashMap, HashSet, VecDeque},
    time::Duration,
};
use tracing::warn;

pub struct DiscoveryConfig<K: KeyType> {
    /// The local peer ID.
    local_peer_id: PeerId,
    /// The local public key.
    local_public_key: PublicKey,
    /// The bootstrap peers.
    bootstrap_peers: Vec<(PeerId, Multiaddr)>,
    /// The relay nodes.
    relay_nodes: Vec<(PeerId, Multiaddr)>,
    /// The number of peers to connect to.
    target_peer_count: u32,
    /// Enable mDNS discovery.
    enable_mdns: bool,
    /// Enable Kademlia discovery.
    enable_kademlia: bool,
    /// Enable `UPnP` discovery.
    enable_upnp: bool,
    /// Enable relay nodes.
    enable_relay: bool,
    /// The name of the network.
    network_name: String,
    /// Protocol version string that uniquely identifies your P2P service.
    protocol_version: String,
    /// Phantom key type
    _marker: blueprint_std::marker::PhantomData<K>,
}

impl<K: KeyType> DiscoveryConfig<K> {
    #[must_use]
    pub fn new(local_public_key: PublicKey, network_name: impl Into<String>) -> Self {
        Self {
            local_peer_id: local_public_key.to_peer_id(),
            local_public_key,
            bootstrap_peers: Vec::new(),
            relay_nodes: Vec::new(),
            target_peer_count: 25, // Reasonable default
            enable_mdns: true,     // Enable by default for local development
            enable_kademlia: true, // Enable by default for production
            enable_upnp: true,     // Enable by default for better connectivity
            enable_relay: true,    // Enable by default for relay functionality
            network_name: network_name.into(),
            protocol_version: String::from("gadget/1.0.0"), // Default version
            _marker: blueprint_std::marker::PhantomData,
        }
    }

    /// Set the protocol version that uniquely identifies your P2P service.
    ///
    /// This should be unique to your application to avoid conflicts with other P2P networks.
    ///
    /// * Format recommendation: `<service-name>/<version>`
    /// * Example: `my-blockchain/1.0.0` or `my-chat-app/0.1.0`
    #[must_use]
    pub fn protocol_version(mut self, version: impl Into<String>) -> Self {
        self.protocol_version = version.into();
        self
    }

    #[must_use]
    pub fn bootstrap_peers(mut self, peers: Vec<(PeerId, Multiaddr)>) -> Self {
        self.bootstrap_peers = peers;
        self
    }

    #[must_use]
    pub fn relay_nodes(mut self, nodes: Vec<(PeerId, Multiaddr)>) -> Self {
        self.relay_nodes = nodes;
        self
    }

    #[must_use]
    pub fn target_peer_count(mut self, count: u32) -> Self {
        self.target_peer_count = count;
        self
    }

    #[must_use]
    pub fn mdns(mut self, enable: bool) -> Self {
        self.enable_mdns = enable;
        self
    }

    #[must_use]
    pub fn kademlia(mut self, enable: bool) -> Self {
        self.enable_kademlia = enable;
        self
    }

    #[must_use]
    pub fn upnp(mut self, enable: bool) -> Self {
        self.enable_upnp = enable;
        self
    }

    #[must_use]
    pub fn relay(mut self, enable: bool) -> Self {
        self.enable_relay = enable;
        self
    }

    /// Construct this [`DiscoveryConfig`] into a [`DiscoveryBehaviour`]
    ///
    /// # Errors
    ///
    /// If `mdns` is enabled, see [`mdns::Behaviour::new`]
    pub fn build(self) -> Result<DiscoveryBehaviour<K>, Error> {
        let kademlia_opt = if self.enable_kademlia {
            let protocol = StreamProtocol::try_from_owned(format!(
                "/gadget/kad/{}/kad/1.0.0",
                self.network_name
            ))?;

            let mut kademlia = new_kademlia(self.local_peer_id, protocol);

            // Add bootstrap peers
            for (peer_id, addr) in &self.bootstrap_peers {
                kademlia.add_address(peer_id, addr.clone());
            }

            // Start bootstrap process
            if let Err(e) = kademlia.bootstrap() {
                warn!("Kademlia bootstrap failed: {}", e);
            }

            Some(kademlia)
        } else {
            None
        };

        let mdns_opt = if self.enable_mdns {
            Some(mdns::Behaviour::new(
                mdns::Config::default(),
                self.local_peer_id,
            )?)
        } else {
            None
        };

        let upnp_opt = if self.enable_upnp {
            Some(upnp::tokio::Behaviour::default())
        } else {
            None
        };

        let relay_opt = if self.enable_relay {
            let relay = relay::Behaviour::new(self.local_peer_id, relay::Config::default());
            Some(relay)
        } else {
            None
        };

        let behaviour = DerivedDiscoveryBehaviour {
            kademlia: kademlia_opt.into(),
            mdns: mdns_opt.into(),
            identify: identify::Behaviour::new(
                identify::Config::new(self.protocol_version, self.local_public_key)
                    .with_agent_version(format!("gadget-{}", env!("CARGO_PKG_VERSION")))
                    .with_push_listen_addr_updates(true),
            ),
            autonat: autonat::Behaviour::new(self.local_peer_id, autonat::Config::default()),
            upnp: upnp_opt.into(),
            relay: relay_opt.into(),
        };

        Ok(DiscoveryBehaviour::<K> {
            discovery: behaviour,
            peers: HashSet::new(),
            peer_info: HashMap::new(),
            target_peer_count: self.target_peer_count,
            next_kad_random_query: tokio::time::interval(Duration::from_secs(1)),
            duration_to_next_kad: Duration::from_secs(1),
            pending_events: VecDeque::new(),
            n_node_connected: 0,
            pending_dial_opts: VecDeque::new(),
            _marker: blueprint_std::marker::PhantomData,
        })
    }
}
