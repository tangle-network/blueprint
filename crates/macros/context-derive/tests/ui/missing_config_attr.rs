use blueprint_sdk::runner::config::BlueprintEnvironment;
use blueprint_context_derive::KeystoreContext;

#[derive(KeystoreContext)]
struct MyContext {
    foo: String,
    sdk_config: BlueprintEnvironment,
}

fn main() {}
