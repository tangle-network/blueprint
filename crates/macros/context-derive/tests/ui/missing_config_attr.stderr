error: No field with #[config] attribute found, please add #[config] to the field that holds the `blueprint_runner::config::BlueprintEnvironment`
 --> tests/ui/missing_config_attr.rs:5:8
  |
5 | struct MyContext {
  |        ^^^^^^^^^
